### PR TITLE
(WIP) Puppet langauge support

### LINKED ISSUE #1281: Puppet Dependency Support
I would like to submit a PR that introduces `Puppetfile` and `metadata.json` support to dependabot. I have working code for Puppetfiles, working on metadata.json next. I realize this is a completely new type of dependency, but hopefully the spec tests, code, and the following background prove this worthwhile. I would like to submit this as a WIP PR, and work to get it into a state to be approved if that is acceptable.

### Puppetfile

A [Puppetfile](https://puppet.com/docs/pe/2019.1/puppetfile.html) lists [Puppet module](https://puppet.com/docs/puppet/6.7/modules_fundamentals.html) dependencies for a Puppet environment.  Each module listed can be from the [Puppet Forge](https://forge.puppet.com/) or from a git reference. There is no lock file for a Puppetfile.

Accepted Puppet Forge module versions are `major.minor.patch` or `:latest` and the resolution strategy is exact match. The general update strategies 'update to latest major' and 'update to latest minor'. There are no ranges supported.

Git references are configurable to use a branch, commit, tag, or sha reference. Their version is evaluated at deployment time not in the Puppetfile. In general, it is not expected for an automated system to be able to 'update' a git reference.

The [Puppetfile](https://puppet.com/docs/pe/2019.1/puppetfile.html) is regarded as Ruby, so it allows data or Ruby code to be included, but is considered a code smell. This means that when it is parsed, it should be in a manner that does not execute untrusted code in the Puppetfile, but just parses the module lists.

### Puppet Modules

A Puppet Module [metatdata.json](https://puppet.com/docs/puppet/6.7/modules_metadata.html) lists required [Puppet modules](https://puppet.com/docs/puppet/6.7/modules_fundamentals.html) that are to be _installed_ with it in order for the module to function. There is no lock file for a metadata.json file.

The resolution strategy is modeled after Ruby gems, in that it supports ranges and exact matches. See the [requirements](https://puppet.com/docs/puppet/6.7/modules_metadata.html#specifying-puppet-version-requirements) section.

### PR DESCRIPTION
**Pull request by jpogran**
This PR introduces support for updating Puppet dependencies. It is still WIP. I expect alot of churn on this PR, and welcome help in how dependabot works.

> Note that this has a commit that adds a VSCode devcontainer to the repo so I could work on this inside a container. I expect this commit to be removed before merge as that's a project level decision and not in scope for this PR

### Provider

- [x] FileFetcher
- [ ] FileParser
- [x] FileUpdater
- [x] MetadataFinder
- [x] Requirement
- [x] Version

### TODO

- [ ] Update Puppetfile parsing to ast not eval
- [ ] Ensure UserAgent used for forgeapi lookups
- [ ] Figure out how to only update minor, or only update patch

### Provider

- [ ] FileFetcher
- [ ] FileParser
- [ ] FileUpdater
- [ ] MetadataFinder
- [ ] Requirement
- [ ] Version

### TODO

- [ ] Everything
- [ ] Figure out ranges ( '~>' and the like )

Fixes #1281

### CODE REVIEW

**Comment by greysteil** [comment]
Sorry it's taking me a while to get to this! Brutally busy week. Will try to review it tonight / tomorrow.

**Comment by greysteil** [comment]
Oh, one more thing: is `Puppetfile.lock` a thing we need to support?

**Comment by jpogran** [comment]
> Oh, one more thing: is `Puppetfile.lock` a thing we need to support?

There aren't lock files for either `Puppetfile` or `metadata.json`, so we're good to leave out that logic.

**Comment by jpogran** [comment]
Thank you so much for the file_update parts, that makes things alot easier and adds alot more functionality.

> It would be great to chat through the approach in the `FileParser`. Can you give me some details on your approach with the `R10K` module?

There's some background in the issue for this PR on Puppetfile format. For our purposes here in a `Puppetfile`, only Puppet Modules and Git references matter as they can be resolved to new versions. Since a `Puppetfile` is ruby, it can either be evaluated or parsed as AST. For the initial POC I copied existing Puppet code in our the Puppet VSCode extension, which was modified from the Puppet r10k project, which uses `instance_eval` to parse the file. Then there are a bunch of helper classes to identify whether its a Puppet module line (something to pull from the Puppet Forge) or from git, etc. This is a fast and reliable approach to parsing the file, but has one large flaw. If there is any other Ruby code in the file, its executed. The better approach is to use an AST, which I have a spike for from a colleague I've yet to push up.

If you want to talk realtime, let me know your preferred method

**Comment by genebean** [comment]
@jpogran - does the logic in use account for Puppetfile entries that are not single line? This page shows some examples https://puppet.com/docs/pe/2017.3/puppetfile.html

**Comment by greysteil** [comment]
Makes sense. We run Dependabot's parsers, update checkers and file updaters in an environment that can handle RCE, but avoiding it is even better, particularly for the parser.

I've looked through everything except the `UpdateChecker` and the `FileParser` now and it's all looking good. I'll take a look through the `UpdateChecker` today / tomorrow.

What do you want to do on the parser? If you push up the AST approach I can take a look (assuming it's  using the same Ruby parser we use in the rest of Dependabot I've got some experience there).

**Comment by jpogran** [comment]
@genebean I think it does, but we can add a test to make sure

**Comment by jpogran** [comment]
@greysteil Thanks for all the work. I'll be able to review this is week, juggling projects at the moment. Should hear from me by 2019/08/14

**Comment by jpogran** [comment]
@greysteil apologies for delay. Here is the AST approach I was referencing: https://github.com/puppetlabs/r10k/pull/885. In short, I think it may be more than we want to take on at this point as it depends on at least ruby v2.6 and as the PR lists there are some implementation concerns.

If you already have methods that deal with Gemfile type evals and have isolated environments, I think that will cover us. I'd rather spend time on finishing out Puppetfile support and then talking about what it would take to add `metadata.json` support.

File: puppet/lib/dependabot/puppet/file_updater/requirement_replacer.rb [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+          end
+
+          def declaration_methods
+            %i(mod)
+          end
+
+          def declares_targeted_module?(node)
+            return false unless declaration_methods.include?(node.children[1])
+
+            node.children[2].children.first == dependency.name.gsub('-', '/')
```
Comment (jpogran) [inline puppet/lib/dependabot/puppet/file_updater/requirement_replacer.rb]: Hi @greysteil I added a commit here that fixes this line to account for the difference in names between the syntax used by Puppetfiles and the naming requirements on the Puppet Forge. In the Puppetfile, there is a `/` splitting the author name and the module name, in the Puppet Forge there is a `-` as a separator. This gsub was in the original set of code, but was lost when we moved to the Ruby AST methods from the other parts of this repo.

While this is the only place that compares the module name, I'm not sure this is the best place to hold this logic. Should this be higher up the chain?

**Comment by jpogran** [comment]
Sorry for the long period of silence @greysteil, summer months and multiple projects shifted priorities. I now have time to see this through, what can I do to help this along?

I re-ran the handlers again using the dry-run and update.rb scripts and found that it created the PR successfully with all the information it had before, but it didn't actually update any files. I found the fix and added a new commit for it. Please check out my comment to see if that is the right place for it

**Comment by genebean** [comment]
Hello @greysteil! I was wondering where things are with this? I am super excited to be able to put it to use.

**Comment by genebean** [comment]
While testing this out tonight I got an entry like below for several modules and don't understand why its saying "no change to Puppetfile" when it shows that there is an update available. Is this expected or a bug in the PR's code?

```
=== ayohrling-local_security_policy (0.6.2)
 => checking for updates
 => requirements to unlock own
 => updating to 0.6.3
    no change to Puppetfile
```

**Comment by rebelagentm** [comment]
@jpogran @genebean Hello! 👋 Grey has since moved off of Dependabot.  The team is currently heads down on scaling Dependabot to all of GitHub, so we're a bit backlogged and slow to respond to issues and PRs.  We will do our best to review the PR from @jpogran and the potential bug @genebean reported as soon as we can.

**Comment by genebean** [comment]
Thanks for the update @rebelagentm

**Comment by genebean** [comment]
@jpogran it seems the git checking of puppet modules is not happening. This can be seen by running the current code against a Puppetfile with the following entry:

```ruby
mod 'autosign', :git => 'git://github.com/danieldreier/puppet-autosign.git', :commit => '17cefbde1c23ecd7e18bc1d3fe8c011d121a278c'
```

That module's out-of-dateness can be seen at https://github.com/danieldreier/puppet-autosign/compare/17cefbde1c23ecd7e18bc1d3fe8c011d121a278c...master

**Comment by genebean** [comment]
@rebelagentm - any chance of this coming back to life soon?

**Comment by rebelagentm** [comment]
@feelepxyz Could you comment on where Dependabot is in terms of reviewing this PR?

**Comment by feelepxyz** [comment]
@jpogran sorry dropping this! Unfortunately can't accept this change as any new package manager makes it harder to debug and maintain for the team which is very under-resourced at the moment 😢

What are your thoughts on building Puppet support as a separate gem, e.g. dependabot-puppet and publishing this yourself? You would have to self-host a runner/scheduler for now but we're investigating ways of allowing you to run dependabot-core using Actions which would improve this situation. You'll want to add a dependency on https://rubygems.org/gems/dependabot-common and use this gemspec as a starting point: https://github.com/dependabot/dependabot-core/blob/master/common/dependabot-common.gemspec

We could also move generic parts to `dependabot-common` if this would help you out over time to make it easier to write custom dependabot-gems. Would love to work towards providing the basic blocks for putting together a language with minimal plumbing, supporting this: #1290

**Pull request closed without merging**